datafusion/common/src/tree_node.rs

```diff
@@ -937,6 +943,28 @@ impl<T> TransformedResult<T> for Result<Transformed<T>> {
     }
 }
 
+/// A node trait which makes all tree traversal iterative and not recursive.
+pub trait RecursiveNode: Sized {
+    /// Read-only access to children
+    fn children(&self) -> Vec<&Self>;
+
+    /// Detaches children from the parent node (if possible).
+    /// Unlike [`ConcreteTreeNode`] it doesn't possible that the value will actually be removed from the parent
+    fn take_children(self) -> (Self, Vec<Self>);
```

Comment:
Could these methods be moved into `TreeNode` itself? It feels like it just makes things additionally complicated to try and implement a separate set of APIs that are related but not quite the same

Reply:
I removed `RecursiveNode` in https://github.com/apache/datafusion/pull/13177/commits/403230aafcad9f75f806b56a2d5c0848b6c3a4c5 by moving its logic into ConcreteTreeNode (though I’m not sure if this could be done directly in TreeNode). 

I also linked `DynTreeNode` and `ConcreteTreeNode`. This change might be debatable, so I’m happy to discuss further if needed :)